File: Codeforces/contest 967 div 2/c.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <set>
#include <utility>
using namespace std;

int query(int a, int b) {
    cout << "? " << a << " " << b << endl;
    cout.flush();
    int res;
    cin >> res;
    return res;
}

void solve(int n) {
    vector<pair<int, int>> edges;
    set<pair<int, int>> added_edges;  // To track added edges and avoid duplicates
    vector<bool> visited(n + 1, false);
    
    queue<int> bfs_queue;
    bfs_queue.push(1);
    visited[1] = true;
    
    while (!bfs_queue.empty()) {
        int current = bfs_queue.front();
        bfs_queue.pop();

        for (int i = 2; i <= n; ++i) {
            if (visited[i]) continue;

            int res = query(current, i);
            
            if (res == current) {
                // Direct connection
                if (!visited[i]) {
                    edges.push_back({current, i});
                    added_edges.insert({current, i});
                    added_edges.insert({i, current});
                    visited[i] = true;
                    bfs_queue.push(i);
                }
            } else {
                // Trace the path from current to i via res
                int tracer = res;
                while (tracer != current) {
                    int next_res = query(current, tracer);
                    if (next_res == current) {
                        if (added_edges.find({current, tracer}) == added_edges.end()) {
                            edges.push_back({current, tracer});
                            added_edges.insert({current, tracer});
                            added_edges.insert({tracer, current});
                        }
                        visited[tracer] = true;
                        bfs_queue.push(tracer);
                        break;
                    } else {
                        tracer = next_res;
                    }
                }
                if (!visited[i]) {
                    if (added_edges.find({tracer, i}) == added_edges.end()) {
                        edges.push_back({tracer, i});
                        added_edges.insert({tracer, i});
                        added_edges.insert({i, tracer});
                    }
                    visited[i] = true;
                    bfs_queue.push(i);
                }
            }
        }
    }

    // Output all unique edges
    cout << "! ";
    for (auto &edge : edges) {
        cout << edge.first << " " << edge.second << " ";
    }
    cout << endl;
    cout.flush();
}

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        solve(n);
    }
    return 0;
}

File: Codeforces/contest 967 div 2/c.py
def q(a, b):
    print(f"? {a} {b}")
    r = int(input())
    return r

def function(a, b, vis, edges):
    x = q(a, b)
    if x == a or x == b:
        edges.append((a, b))
        vis[x] = 1
        vis[b] = 1
        return
    if vis[a] and vis[x]:
        function(x, b, vis, edges)
    else:
        function(a, x, vis, edges)
        function(x, b, vis, edges)

def solve():
    t = int(input())
    while t > 0:
        t -= 1
        n = int(input())

        if n == 2:
            print("! 1 2")
            continue

        vis = [0] * (n + 1)
        edges = []

        for i in range(1, n):
            for j in range(i + 1, n + 1):
                if vis[j]:
                    continue
                function(i, j, vis, edges)

        print("!", end="")
        for edge in edges:
            print(f" {edge[0]} {edge[1]}", end="")
        print()

solve()
